ocaml/testsuite/tests/parsetree/test.ml

```diff
@@ -79,31 +87,61 @@ let test parse_fun pprint print map filename =
             print_endline"====================================================="
           end
 
-let test parse_fun pprint print map filename =
-  try test parse_fun pprint print map filename
+let test parse_fun pprint print map filename ~extra_checks =
+  try test parse_fun pprint print map filename ~extra_checks
   with exn -> report_err exn
 
-let rec process path =
+let rec process path ~extra_checks =
   if Sys.is_directory path then
     let files = Sys.readdir path in
-    Array.iter (fun s -> process (Filename.concat path s)) files
+    Array.iter (fun s -> process (Filename.concat path s) ~extra_checks) files
   else if Filename.check_suffix path ".ml" then
     test
       Parse.implementation
       Pprintast.structure
       Printast.implementation
       (fun mapper -> mapper.Ast_mapper.structure)
       path
+      ~extra_checks
   else if Filename.check_suffix path ".mli" then
     test
       Parse.interface
       Pprintast.signature
       Printast.interface
       (fun mapper -> mapper.Ast_mapper.signature)
       path
+      ~extra_checks
+
+let process ?(extra_checks = fun _ -> Ok ()) text = process text ~extra_checks
+
+let check_all_extension_points_start_with text ~prefix =
+  let exception Unexpected_extension_point of string in
+  let check ~extension_point_prefix =
+    String.split_on_char '[' text
+    |> List.for_all (fun s ->
+        not (String.starts_with s ~prefix:extension_point_prefix)
+        ||   String.starts_with s ~prefix:(extension_point_prefix ^ prefix))
+  in
```

Comment:
Some minor notes:
1. Maybe "extension point" isn't the best name, since "extension _node_" specifically means `[%foo]`.  Maybe just `introduction_character`?
2. This can miss `[@@test.foo]`, `[%%test.bar]`, and `[@@@test.baz]`.
3. This can miss `[@ test.foo]` (note the space).
Now, the latter two will only result in false positive test failures, so it's not necessarily a huge deal, but if we don't fix it, we should document it.

I also found the conditional confusing; I really wish more languages had a "boolean implication operator".  `String.starts_with s ~prefix:extension_point_prefix) ==> String.starts_with s ~prefix:(extension_point_prefix ^ prefix))` would be substantially clearer. /gripe

Reply:
1. Fair enough, I inlined "attribute and extension node" in more places.
2. This mishandles `[@@` in the conservative direction. Luckily, the test doesn't have any instances of `@@` or `%%` (etc.) so we can just punt on this.
3. Same as (2).

`a ==> b` is close enough to `if a then b else true`, so I did that instead.

Reply:
I agree re. conservativity, and it seems hard to get right -- I'd just like to add a comment

Reply:
I've added that. Can you approve if you think it looks good?